File: tests/regression/benchmark.r
# bench_multiple_sizes.R
bench_r <- function(n, p, trials=5) {
  X <- matrix(rnorm(n*p), n, p)
  y <- rnorm(n)
  
  # Warmup
  fit <- lm.fit(X, y)
  
  times <- numeric(trials)
  for (i in 1:trials) {
    times[i] <- system.time(fit <- lm.fit(X, y))["elapsed"]
  }
  
  median(times)
}

cat("R lm.fit() Benchmark\n")
cat(paste0(rep("=", 60), collapse=""), "\n")

for (config in list(c(1000, 10), c(10000, 50), c(100000, 100), 
                     c(500000, 200), c(1000000, 500))) {
  n <- config[1]
  p <- config[2]
  t <- bench_r(n, p)
  cat(sprintf("n=%7d, p=%4d: %8.2f ms\n", n, p, t * 1000))
}

File: tests/regression/benchmark.py
# bench_multiple_sizes.py
import numpy as np
import time
from pystatistics.regression.design import Design
from pystatistics.regression.backends.cpu import CPUQRBackend

def bench_pystatistics(n, p, trials=5):
    X = np.random.randn(n, p)
    y = np.random.randn(n)
    
    design = Design.from_arrays(X=X, y=y)
    backend = CPUQRBackend()
    
    # Warmup
    _ = backend.solve(design)
    
    times = []
    for _ in range(trials):
        start = time.perf_counter()
        result = backend.solve(design)
        times.append(time.perf_counter() - start)
    
    return np.median(times)

print("PyStatistics CPU Benchmark")
print("="*60)
for n, p in [(1000, 10), (10000, 50), (100000, 100), (500000, 200), (1000000, 500)]:
    t = bench_pystatistics(n, p)
    print(f"n={n:>7}, p={p:>4}: {t*1000:>8.2f} ms")